x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceStatsFilteredAggWithEval.java

```diff
@@ -31,44 +34,99 @@
-                Alias newAlias = alias.replaceChild(Literal.of(aggFunction, value));
-                newEvals.add(newAlias);
-                newProjections.add(newAlias.toAttribute());
-            } else {
-                newAggs.add(ne); // agg function unchanged or grouping key
-                newProjections.add(ne.toAttribute());
+                    Object value = aggFunction instanceof Count || aggFunction instanceof CountDistinct ? 0L : null;
+                    Alias newAlias = alias.replaceChild(Literal.of(aggFunction, value));
+                    newEvals.add(newAlias);
+                    newProjections.add(newAlias.toAttribute());
+                } else {
+                    newAggs.add(ne); // agg function unchanged or grouping key
+                    newProjections.add(ne.toAttribute());
+                }
             }
-        }
 
-        LogicalPlan plan = aggregate;
-        if (newEvals.isEmpty() == false) {
-            if (newAggs.isEmpty()) { // the Aggregate node is pruned
-                plan = localRelation(aggregate.source(), newEvals);
-            } else {
-                plan = aggregate.with(aggregate.child(), aggregate.groupings(), newAggs);
-                plan = new Eval(aggregate.source(), plan, newEvals);
-                plan = new Project(aggregate.source(), plan, newProjections);
+            if (newEvals.isEmpty() == false) {
+                if (newAggs.isEmpty()) { // the Aggregate node is pruned
+                    if (ij.get() != null) { // this is an Aggregate part of right-hand side of an InlineJoin
+                        // replace the right hand side Aggregate with an Eval
+                        InlineJoin newIJ = new InlineJoin(ij.get().source(), ij.get().left(), ij.get().right().transformDown(p -> {
+                            // both following steps are executed for an InlineJoin
+                            if (p instanceof Aggregate a && a == aggregate) {
+                                // the aggregate becomes a simple Eval since it's not needed anymore (it was replaced with Literals)
+                                p = new Eval(aggregate.source(), aggregate.child(), newEvals);
+                            } else if (p instanceof StubRelation) {
+                                // Remove the StubRelation since the right-hand side of the join is now part of the main plan
+                                // and it won't be executed separately by the EsqlSession inlinestats planning.
+                                p = ij.get().left();
+                            }
+                            return p;
+                        }), ij.get().config());
+                        // project the correct output (the one of the former inlinejoin) and remove the InlineJoin altogether,
+                        // replacing it with its right-hand side followed by its left-hand side
+                        plan = new Project(newIJ.source(), newIJ.right(), newIJ.output());
```

Comment:
This `Project` is what's basically returned. Can't the entire `InlineJoin` transformation be skipped and only transform with the right hand-side of the "old" `InlineJoin`?